### DISCUSSION TITLE
quarkus-scheduler: Timestamps of the current run and the previous run are indistinguishable

### DISCUSSION THREAD

**Discussion by JBartscher** [quarkusio/quarkus#40789] [Q&A]
Hi all,

I have noticed a peculiarity in the extension quarkus-scheduler that I am not sure if it is a bug or if I am simply using the lib incorrectly. 
The issue is that the quarkus-scheduler uses the ScheduledExecution class to map the context of a task scheduled by the scheduler. This allows the execution time to be determined using the `getScheduledFireTime()` function. The trigger can also be accessed using `getTrigger()`. This in turn offers the function `getPreviousFireTime()` to terminate the last run of the task. 

Now the curious thing: these two timestamps are identical.
![image](https://github.com/quarkusio/quarkus/assets/12815584/e88446c0-f50d-485b-8a56-5d0ae0758419)

Am I using the scheduler and its methods incorrectly or is this an actual bug?

Ive set up a minimal example to test this behavior if anyone want to have a look at it.
https://github.com/JBartscher/scheduler-ctx-timestamp-bug 

Ive tested it with Quarkus in Version 3.10.1

[a comment by an automated account is not shown]

**Comment by manovotn** [comment]
I found a very similar Quarkus issue that might be of interest to you - https://github.com/quarkusio/quarkus/issues/37567

@mkouba's answer there says:

> The reason why the `Trigger#getPreviousFireTime()` is sometimes the same as `ScheduledExecution.getFireTime()` is that trigger is a mutable object that reflects the current state whereas the `ScheduledExecution` is immutable. In other words, `ScheduledExecution.getFireTime()` always returns the same value whereas `Trigger#getPreviousFireTime()` returns the actual value.

I also tried looking into Quartz's API which have very similar methods (see [here](https://www.quartz-scheduler.org/api/2.3.0/org/quartz/JobExecutionContext.html#getPreviousFireTime--)) but nonexistent javadoc to explain what the diff should be :-/

**Reply by mkouba** [reply]
Yes, so the difference between `Trigger` and `ScheduledExecution` is that the trigger is bound to a scheduled method/job whereas the scheduled execution is bound to a particular invocation of `@Scheduled` method or job registered programmatically with `Scheduler.newJob()`. 

The trigger is a stateful object, that is used to test if a scheduled job should be executed. In other words, both `Trigger#getNextFireTime()` and `Trigger#getPreviousFireTime()` may return different values in time. 

The `ScheduledExecution`, on the other hand, is an immutable snapshot and both `ScheduledExecution#getFireTime()`and `ScheduledExecution#getScheduledFireTime()` always returns the same value.

Note that you can also obtain the trigger with the `Scheduler#getScheduledJob(String)` method, i.e. outside a specific scheduled execution.
